garden_ai/model.py

```diff
@@ -0,0 +1,146 @@
+        "en", "de", "fr". Defaults to "en".
+
+    subjects: List[str]
+        Subjects, keywords, classification codes, or key phrases pertaining to the Garden.
+
+    Examples
+    --------
+    Creating a new Garden with a ``GardenClient`` instance::
+
+        gc = GardenClient()
+        pea_garden = gc.create_garden(
+            creators=["Mendel, Gregor"], title="Experiments on Plant Hybridization"
+        )
+        pea_garden.creators += ["St Thomas' Abbey"]
+        pea_garden.year = 1863
+        pea_garden.description = '''This Garden houses sophisticated ML pipelines
+                                  for Big Pea Data extraction and classification.
+                                  It consists of a 2-hectare plot behind the monastery,
+                                  and a 30,000-plant dataset.'''
+
+        gc.register(pea_garden) # we can also have this invoke a pydantic "root validator" we define elsewhere
+
+    Notes
+    --------
+    Mendel's work was ignored by the scientific community during his lifetime,
+    presumably due to the lack of a working DOI.
+    To remedy this, if the doi field is unset when registering the
+    garden, we could just build one for the user with the datacite api
+    """
+
+    class Config:
+        """
+        Configure pydantic per-model settings.
+
+        We disable validate_all so that pydantic ignores our custom TempField defaults
+        We enable validate_assignment to make validation occur naturally even after object construction
+        """
+
+        validate_all = False
+        validate_assignment = True  # validators invoked (only?) on assignment
+
+    creators: List[str] = Field(TempField.REQUIRED, min_items=1, unique_items=True)
+    title: str = Field(TempField.REQUIRED)
+    resourceTypeGeneral: str = "Other"  # (or: model, software, service, interactive?)
+    publisher: str = "Garden"
+    year: str = Field(default_factory=lambda: str(datetime.now().year))
+
+    # dois currently only validated against this regex
+    # doi: str = Field(TempField.REQUIRED, alias="identifier", regex=r"^10.\d{4,9}/[-._;()/:A-Z0-9]+$")
+    doi: str = Field(TempField.REQUIRED)
+
+    language: str = "en"
+    # subjects: List[str] = Field(TempField.WARN, unique_items=True)
+    version: str = Field(TempField.WARN)
+    description: str = Field(TempField.WARN)
+
+    # field(s) for which we might want to ''disable'' mutation
+    garden_id: UUID = Field(default_factory=uuid4, allow_mutation=False)
+
+    # always=False should prevent this from being called until an attempt to set
+    # a field, but this should be called for every field.
+    @validator("*", pre=True, always=False)
+    def check_required(cls, val, field: pydantic.fields.ModelField):
+        assert val is not TempField.REQUIRED, f"{field.name} must be set"
+        if val is TempField.WARN:
+            # TODO: log warning
+            print(f"Warning: {field.name} is recommended but unset")
```

Comment:
Wouldn't it make more sense to do logging instead of print statements?

Reply:
yes, good idea Aadit!